### PR TITLE
[SPARK-47273][SS][PYTHON] implement Python data stream writer interface.

### PR DESCRIPTION
**Pull request by chaoqin-li1123**
### What changes were proposed in this pull request?
Reuse PythonPartitionWriter to implement the serialization and execution of write callback in executor.

Implement python worker process to run python streaming data sink committer and communicate with JVM through socket in spark driver. For each python streaming data sink instance there will be a long live python worker process created. Inside the python process, the python write committer will receive abort or commit function call and send back result through socket.

### Why are the changes needed?
In order to support developing spark streaming sink in python, we need to implement python stream writer interface.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Unit and integration test.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by chaoqin-li1123** [comment]
PTAL,Thanks! @HeartSaVioR @HyukjinKwon @allisonwang-db @sahnib

**Comment by HyukjinKwon** [comment]
Let's also fix up the linter failure.

File: python/pyspark/sql/datasource.py [hunk 1]
```diff
@@ -516,6 +528,71 @@ def abort(self, messages: List["WriterCommitMessage"]) -> None:
...
+        each microbatch. It accepts an iterator of input data and returns a single row
+        representing a commit message, or None if there is no commit message.
+
+        The driver collects commit messages, if any, from all executors and passes them
+        to the ``commit`` method if all tasks run successfully. If any task fails, the
+        ``abort`` method will be called with the collected commit messages.
+
+        Parameters
+        ----------
+        iterator : Iterator[Row]
```
Comment (HyukjinKwon) [inline python/pyspark/sql/datasource.py:554]: Would you mind creating a PR to fix those all up after several related PRs merged? It really has to be something like `` iterator of :class:`Row` ``

File: python/pyspark/sql/datasource.py [hunk 2]
```diff
@@ -516,6 +528,71 @@ def abort(self, messages: List["WriterCommitMessage"]) -> None:
...
+            An iterator of input data.
+
+        Returns
+        -------
+        WriterCommitMessage : a serializable commit message
```
Comment (HyukjinKwon) [inline python/pyspark/sql/datasource.py:559]: This also should really be `` :class:`WriterCommitMessage` ``

File: python/pyspark/sql/datasource.py [hunk 3]
```diff
@@ -516,6 +528,71 @@ def abort(self, messages: List["WriterCommitMessage"]) -> None:
...
+        Aborts this microbatch due to task failures.
+
+        This method is invoked on the driver when one or more tasks failed. The commit
+        messages are collected from the ``write`` method call from each task, and are
+        passed to this method. The implementation should use the commit messages to
+        abort the microbatch in the streaming sink.
+
+        Parameters
+        ----------
+        messages : List[WriterCommitMessage]
```
Comment (HyukjinKwon) [inline python/pyspark/sql/datasource.py:590]: `` list of :class:`WriterCommitMessage` ``

File: python/pyspark/sql/datasource.py [hunk 4]
```diff
@@ -160,6 +160,18 @@ def writer(self, schema: StructType, overwrite: bool) -> "DataSourceWriter":
             message_parameters={"feature": "writer"},
         )
 
+    def streamWriter(self, schema: StructType, overwrite: bool) -> "DataSourceStreamWriter":
+        """
+
+        :param schema:
+        :param overwrite:
+        :return:
```
Comment (HyukjinKwon) [inline python/pyspark/sql/datasource.py:168]: docs?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/python/PythonStreamingDataSourceSuite.scala [hunk 5]
```diff
@@ -16,15 +16,24 @@
...
 import org.apache.spark.sql.execution.datasources.v2.python.{PythonDataSourceV2, PythonMicroBatchStream, PythonStreamingSourceOffset}
+import org.apache.spark.sql.execution.streaming.MemoryStream
 import org.apache.spark.sql.types.StructType
 import org.apache.spark.sql.util.CaseInsensitiveStringMap
 
 class PythonStreamingDataSourceSuite extends PythonDataSourceSuiteBase {
 
+  import testImplicits._
+
+  val waitTimeout = 5.seconds
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/python/PythonStreamingDataSourceSuite.scala:35]: are 5 secs enough? I would set it to 10 or 30 secs to be sure. Streaming with Python is is pretty flaky and slow time to time.

**Review by HyukjinKwon — APPROVED** [review]
Seems fine, cc @allisonwang-db

File: sql/core/src/test/scala/org/apache/spark/sql/execution/python/PythonStreamingDataSourceSuite.scala (hunk 5 above)
Comment (chaoqin-li1123) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/python/PythonStreamingDataSourceSuite.scala]: Changed to 15 seconds.

File: python/pyspark/sql/datasource.py (hunk 4 above)
Comment (chaoqin-li1123) [reply python/pyspark/sql/datasource.py]: Docs added.

File: python/pyspark/sql/datasource.py [hunk 6]
```diff
@@ -513,6 +536,71 @@ def abort(self, messages: List["WriterCommitMessage"]) -> None:
...
+        This method is invoked on the driver when all tasks run successfully. The
+        commit messages are collected from the ``write`` method call from each task,
+        and are passed to this method. The implementation should use the commit messages
+        to commit the microbatch in the streaming sink.
+
+        Parameters
+        ----------
+        messages : List[WriterCommitMessage]
+            A list of commit messages.
+        """
```
Comment (allisonwang-db) [inline python/pyspark/sql/datasource.py:584]: how about `batchId`

File: python/pyspark/sql/datasource.py [hunk 7]
```diff
@@ -513,6 +536,71 @@ def abort(self, messages: List["WriterCommitMessage"]) -> None:
...
+
+        This method is invoked on the driver when one or more tasks failed. The commit
+        messages are collected from the ``write`` method call from each task, and are
+        passed to this method. The implementation should use the commit messages to
+        abort the microbatch in the streaming sink.
+
+        Parameters
+        ----------
+        messages : List[WriterCommitMessage]
+            A list of commit messages.
```
Comment (allisonwang-db) [inline python/pyspark/sql/datasource.py:599]: ditto for `batchId`

File: python/pyspark/sql/worker/python_streaming_sink_runner.py [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+    utf8_deserializer,
+)
+
+
+def main(infile: IO, outfile: IO) -> None:
+    try:
+        check_python_version(infile)
+        setup_spark_files(infile)
+
+        memory_limit_mb = int(os.environ.get("PYSPARK_PLANNER_MEMORY_MB", "-1"))
```
Comment (allisonwang-db) [inline python/pyspark/sql/worker/python_streaming_sink_runner.py:54]: nit: you might want to use a different config for streaming workers.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/python/PythonStreamingSinkCommitRunner.scala [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+import org.apache.spark.SparkEnv
+import org.apache.spark.api.python.{PythonFunction, PythonWorker, PythonWorkerFactory, PythonWorkerUtils, SpecialLengths}
+import org.apache.spark.internal.Logging
+import org.apache.spark.internal.config.BUFFER_SIZE
+import org.apache.spark.internal.config.Python.PYTHON_AUTH_SOCKET_TIMEOUT
+import org.apache.spark.sql.connector.write.WriterCommitMessage
+import org.apache.spark.sql.errors.QueryExecutionErrors
+import org.apache.spark.sql.types.StructType
+
+class PythonStreamingSinkCommitRunner(
```
Comment (allisonwang-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/python/PythonStreamingSinkCommitRunner.scala:33]: Can we add a scaladoc for this class?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/python/PythonStreamingWrite.scala [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+ */
+package org.apache.spark.sql.execution.datasources.v2.python
+
+import org.apache.spark.JobArtifactSet
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.connector.write._
+import org.apache.spark.sql.connector.write.streaming.{StreamingDataWriterFactory, StreamingWrite}
+import org.apache.spark.sql.types.StructType
+
+class PythonStreamingWrite(
```
Comment (allisonwang-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/python/PythonStreamingWrite.scala:25]: ditto for scaladoc.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/python/PythonStreamingDataSourceSuite.scala [hunk 11]
```diff
@@ -230,4 +273,220 @@ class PythonStreamingDataSourceSuite extends PythonDataSourceSuiteBase {
...
+      withTempDir { dir =>
+        val path = dir.getAbsolutePath
+        val checkpointDir = new File(path, "checkpoint")
+        checkpointDir.mkdir()
+        val outputDir = new File(path, "output")
+        outputDir.mkdir()
+        val streamDF = if (mode == "append") {
+          inputData.toDF()
+        } else {
+          // Complete mode only supports stateful aggregation
```
Comment (allisonwang-db) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/python/PythonStreamingDataSourceSuite.scala:293]: Forgive my limited knowledge of streaming, but what does 'complete' mode mean here? Is it the same as setting 'overwrite=true'?

**Review by allisonwang-db — COMMENTED** [review]
Looks good overall!

File: sql/core/src/test/scala/org/apache/spark/sql/execution/python/PythonStreamingDataSourceSuite.scala (hunk 11 above)
Comment (chaoqin-li1123) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/python/PythonStreamingDataSourceSuite.scala:293]: My understanding is that Complete mode is equivalent to 'overwrite=true'. Correct me if I am wrong. @HeartSaVioR

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/python/PythonStreamingSinkCommitRunner.scala (hunk 9 above)
Comment (chaoqin-li1123) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/python/PythonStreamingSinkCommitRunner.scala:40]: scala doc added.

File: python/pyspark/sql/worker/python_streaming_sink_runner.py (hunk 8 above)
Comment (chaoqin-li1123) [reply python/pyspark/sql/worker/python_streaming_sink_runner.py:54]: The sink commit runner is relatively lightweight, so reusing planner memory limit should be fine.

File: python/pyspark/sql/datasource.py (hunk 6 above)
Comment (chaoqin-li1123) [reply python/pyspark/sql/datasource.py]: Doc for batchId added.

**Review by allisonwang-db — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/execution/python/PythonStreamingDataSourceSuite.scala (hunk 11 above)
Comment (HeartSaVioR) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/python/PythonStreamingDataSourceSuite.scala:293]: complete mode requires sink to support truncate.

**Review by HeartSaVioR — APPROVED** [review]
+1

File: python/pyspark/sql/datasource.py [hunk 12]
```diff
@@ -160,6 +160,29 @@ def writer(self, schema: StructType, overwrite: bool) -> "DataSourceWriter":
             message_parameters={"feature": "writer"},
         )
 
+    def streamWriter(self, schema: StructType, overwrite: bool) -> "DataSourceStreamWriter":
```
[the remark this answers is not in the document]
Comment (HeartSaVioR) [reply python/pyspark/sql/datasource.py:163]: I see it's for truncate. Never mind.

**Comment by HeartSaVioR** [comment]
I expect the "python" end-to-end tests to be added later. Other than that, we can move forward.

Thanks! Merging to master.

**Pull request closed without merging**